### PR TITLE
HPCC-19586 Prevent moving filters over joins if it creates an ambiguity

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review.  Example that illustrates the problem is janderson1.eclxml.  Others are also affected (and may have been wrong before) e.g. srao2, aslow4, salini4

**Comment by ghalliday** [comment]
(Easiest to compare with ?w=1 to ignore whitespace)

File: ecl/hql/hqlopt.cpp [hunk 1]
```diff
@@ -1166,51 +1166,61 @@ IHqlExpression * CTreeOptimizer::getHoistedFilter(IHqlExpression * transformed,
             }
         }
 
-        if (!matched && !expandMonitor.isComplex())
+        try
```
Comment (shamser) [inline ecl/hql/hqlopt.cpp]: It's a little more efficient to have the try block inside the if block below.

**Review by shamser — COMMENTED** [review]
@ghalliday Reviewed.  One minor comment.

**Review by shamser — COMMENTED** [review]
@ghalliday

File: ecl/hql/hqlopt.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlopt.cpp]: try blocks have no cost with modern implementations - so there is no difference in efficiency.

**Comment by ghalliday** [comment]
@shamser see comment above.  @richardkchapman please review.

**Pull request merged**